File: backend/snooper/src/main/java/com/github/vasniktel/snooper/backend/subscriptions/SubscriptionRepository.java
package com.github.vasniktel.snooper.backend.subscriptions;

import org.springframework.data.repository.CrudRepository;
import org.springframework.transaction.annotation.Transactional;

public interface SubscriptionRepository extends CrudRepository<SubscriptionDto, Integer> {
  SubscriptionDto getFirstByFollowerIdAndFolloweeId(String followerId, String followeeId);

  @Transactional
  void deleteByFollowerIdAndFolloweeId(String followerId, String followeeId);
}

File: app/src/main/java/com/github/vasniktel/snooper/logic/subscription/SubscriptionRepository.kt
package com.github.vasniktel.snooper.logic.subscription

interface SubscriptionRepository {
    suspend fun add(followerId: String, followeeId: String)
    suspend fun remove(followerId: String, followeeId: String)
    suspend fun isFollowee(followerId: String, followeeId: String): Boolean
}

class SubscriptionRepositoryImpl(
    private val remote: RemoteSubscriptionDataSource
) : SubscriptionRepository {
    override suspend fun add(followerId: String, followeeId: String) {
        remote.add(followerId, followeeId)
    }

    override suspend fun remove(followerId: String, followeeId: String) {
        remote.remove(followerId, followeeId)
    }

    override suspend fun isFollowee(followerId: String, followeeId: String): Boolean {
        return remote.isFollowee(followerId, followeeId)
    }
}
